nav2_collision_monitor/src/polygon_source.cpp

```diff
@@ -0,0 +1,215 @@
+      // frames.
+      if (
+        !nav2_util::getTransform(
+          polygon_stamped.header.frame_id, base_frame_id_,
+          transform_tolerance_, tf_buffer_, tf_transform))
+      {
+        return;
+      }
+    }
+    geometry_msgs::msg::PolygonStamped poly_out;
+    geometry_msgs::msg::TransformStamped tf = tf2::toMsg(tf_transform);
+    tf2::doTransform(polygon_stamped, poly_out, tf);
+    convertPolygonStampedToPoints(poly_out, data);
+  }
+}
+
+void PolygonSource::convertPolygonStampedToPoints(
+  const geometry_msgs::msg::PolygonStamped & polygon, std::vector<Point> & data) const
+{
+  /* Generated by GPT */
+
+  // Iterate over the vertices of the polygon
+  for (size_t i = 0; i < polygon.polygon.points.size(); ++i) {
+    const auto & current_point = polygon.polygon.points[i];
+    const auto & next_point = polygon.polygon.points[(i + 1) % polygon.polygon.points.size()];
+
+    // Calculate the distance between the current and next points
+    double segment_length =
+      std::hypot(next_point.x - current_point.x, next_point.y - current_point.y);
+
+    // Calculate the number of points to sample in the current segment
+    int num_points_in_segment =
+      std::max(static_cast<int>(std::ceil(segment_length / sampling_distance_)), 1);
+
+    // Calculate the step size for each pair of vertices
+    const double dx = (next_point.x - current_point.x) / num_points_in_segment;
+    const double dy = (next_point.y - current_point.y) / num_points_in_segment;
+
+    // Sample the points with equal spacing
+    for (int j = 0; j <= num_points_in_segment; ++j) {
+      Point p;
+      p.x = current_point.x + j * dx;
+      p.y = current_point.y + j * dy;
+      data.push_back(p);
+    }
+  }
+}
+
+void PolygonSource::getParameters(std::string & source_topic)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+
+  getCommonParameters(source_topic);
+
+  nav2_util::declare_parameter_if_not_declared(
+    node, source_name_ + ".sampling_distance", rclcpp::ParameterValue(0.1));
+  sampling_distance_ = node->get_parameter(source_name_ + ".sampling_distance").as_double();
+  nav2_util::declare_parameter_if_not_declared(
+    node, source_name_ + ".polygon_similarity_threshold", rclcpp::ParameterValue(2.0));
+  polygon_similarity_threshold_ =
+    node->get_parameter(source_name_ + ".polygon_similarity_threshold").as_double();
+}
+
+void PolygonSource::dataCallback(geometry_msgs::msg::PolygonStamped::ConstSharedPtr msg)
+{
+  auto node = node_.lock();
+  if (!node) {
+    throw std::runtime_error{"Failed to lock node"};
+  }
+  auto curr_time = node->now();
+
+  // remove old data
+  data_.erase(
+    std::remove_if(
+      data_.begin(), data_.end(),
+      [this, curr_time](const geometry_msgs::msg::PolygonStamped & polygon_stamped) {
+        return curr_time - rclcpp::Time(polygon_stamped.header.stamp) > source_timeout_;
```

Comment:
I can see why you thought you could remove the check with source timeout from the main file -- but there's no promise that this callback is run in any particular frequency or regularity. We still need to check the polygon's timeout in the main getData function

Reply:
Better yet, move this up there -- we're looping through the data anyway in `getData`, do the deletion there and remove here

Reply:
You're right that the removal of outdated polygons can't be done here; I had it in the getData function originally but the problem is that the getData function is marked as `const`, meaning I can't modify `data_` in it. Are you fine with removing the const from that function? Keep in mind it has to be removed for all sources.

Reply:
Yeah, I'm fine with that!